python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+    """
+    batch_buffer: List[ObjectRef[Block]] = []
+    buffer_size = 0
+    starting_index = 0
+
+    global_index = 0
+    num_rows_in_last_block = 0
+
+    if batch_size is None:
+        for block_ref, metadata in block_ref_iterator:
+            yield LogicalBatch(global_index, [block_ref], 0, None, metadata.num_rows)
+            global_index += 1
+    else:
+        while True:
+            if buffer_size < batch_size:
+                # Pull next block from iterator if current buffer is not enough to fill
+                # a batch.
+                try:
+                    block_ref, metadata = next(block_ref_iterator)
+                except StopIteration:
+                    break
+                batch_buffer.append(block_ref)
+                buffer_size += metadata.num_rows
+                num_rows_in_last_block = metadata.num_rows
+
+            if buffer_size == batch_size:
+                # If equal to batch size, then yield the full buffer.
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, None, buffer_size
+                )
+                batch_buffer = []
+                buffer_size = 0
+                starting_index = 0
+                num_rows_in_last_block = 0
+                global_index += 1
+
+            if buffer_size > batch_size:
+                # If current buffer is greater than batch size, then yield part of the
+                # buffer, and carryover the remainder to the next batch.
+                num_rows_to_leave_behind = buffer_size - batch_size
+                ending_index = num_rows_in_last_block - num_rows_to_leave_behind
+                assert ending_index > 0, ending_index
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, ending_index, batch_size
+                )
+                global_index += 1
+                # Carryover to next batch.
+                batch_buffer = [batch_buffer[-1]]
+                buffer_size = num_rows_to_leave_behind
+                starting_index = ending_index
+
+    # Yield any leftover batches if necessary.
+    if buffer_size > 0 and not drop_last:
+        assert buffer_size < batch_size
+        yield LogicalBatch(
+            global_index, batch_buffer, starting_index, None, buffer_size
+        )
+        global_index += 1
+
+
+def local_shuffle_logical_batches(
+    logical_batch_iterator: Iterator[LogicalBatch],
+    shuffle_buffer_min_size: int,
+    shuffle_seed: Optional[int] = None,
+) -> Iterator[LogicalBatch]:
+    """Shuffles the logical batch iterator using a buffer of the provided size."""
+
+    if shuffle_seed is not None:
+        random.seed(shuffle_seed)
+
+    shuffle_buffer: List[LogicalBatch] = []
+    shuffle_buffer_size = 0
+    global_counter = 0
+
+    for logical_batch in logical_batch_iterator:
+        shuffle_buffer.append(logical_batch)
```

Comment:
Doesn't this only shuffle logical batches and not the rows within the logical batches like the original impl is supposed to?

Reply:
hmm yeah that's right....shuffling with rows is tricky since we can't do it in the threadpool, but needs to be done before collations

Reply:
We could lock around a shared shuffle buffer data structure here... it should be reasonable enough unless shuffle itself becomes the bottleneck, which is probably unlikely.

Reply:
Yeah this does need to happen after object resolution in order to preserve the existing local shuffling semantics, which I think we want to do.

I'd vote for locking around the existing `ShufflingBatcher` abstraction or some sensible port, which is already optimized to efficiently interleave additions to the buffer and popping of shuffled batches: it will cheaply add blocks to the buffer until the minimum buffer size is met and the consumer requests a batch, it will then generate shuffle indices once for the enter buffer, and yield shuffled batches via those shuffle indices, making runs of block additions and runs of batch pops efficient.

This is a natural merge/sync point for the data-parallel threadpool anyway, since we need to fill the shuffle buffer as fast as possible so we can meet the min shuffle buffer size threshold ASAP, and the shuffle operation needs to be global in terms of the shuffle buffer, so you wouldn't want to silo the data in multiple shuffle buffers on different threads. We'd still be able to apply formatting in parallel on the shuffled batches.

Reply:
Updated to using `ShufflingBatcher` directly (which will be done in a single thread)